ledger-core/network/nds/msgdelivery/ctl_retry_msg_worker.go

```diff
@@ -106,78 +121,71 @@ func (p *retryMsgWorker) sendHead(msg *msgShipment, repeatFn func(retries.RetryI
 func noRetryFn(retries.RetryID) {}
 
 func (p *retryMsgWorker) processOoB(msg *msgShipment) {
-	var repeatFn func(retries.RetryID)
-
 	if msg.isFireAndForget() {
-		repeatFn = noRetryFn
+		p.processMsg(msg, noRetryFn)
 	} else {
-		repeatFn = p.sender.stages.AddHeadForRetry
+		p.processMsg(msg, p.sender.stages.AddHeadForRetry)
 	}
+}
 
-	switch {
-	case p.marks.mark(msg.id):
-		p.sema.Lock()
-		go p.sendHead(msg, repeatFn)
-	case msg.canSendHead():
-		p.pushPostponed(msg, repeatFn)
+func (p *retryMsgWorker) processMsg(msg *msgShipment, repeatFn func(retries.RetryID)) bool {
+	if msg == nil {
+		return true
 	}
-}
 
-func (p *retryMsgWorker) processMsg(msg *msgShipment, repeatFn func(retries.RetryID)) {
+	captured, postponer := p.marks.mark(msg.id, p)
 	switch {
-	case msg == nil:
-		return
-	case !p.marks.mark(msg.id):
-		//
+	case !captured:
 	case !p.sema.TryLock():
 		p.marks.unmark(msg.id)
 	default:
 		go p.sendHead(msg, repeatFn)
-		return
+		return true
 	}
 
 	if msg.canSendHead() {
-		p.pushPostponed(msg, repeatFn)
+		postponer.pushPostponed(msg, repeatFn)
+		return false
 	}
+	return true
 }
 
 func (p *retryMsgWorker) pushPostponed(msg *msgShipment, repeatFn func(retries.RetryID)) {
 	if msg == nil {
 		return
 	}
-	if prev := p.postponed[p.write]; prev.msg != nil {
-		if prev.msg.canSendHead() && prev.repeatFn != nil {
-			prev.repeatFn(retries.RetryID(prev.msg.id))
+
+	select {
+	case p.postponed <- postponedMsg{msg, repeatFn}:
+		return
+	default:
+		if msg.canSendHead() && repeatFn != nil {
+			repeatFn(retries.RetryID(msg.id))
 		} else {
-			prev.msg.markCancel()
+			msg.markCancel()
 		}
 	}
-	p.postponed[p.write] = postponedMsg{msg, repeatFn}
-	if p.write++; p.write >= int32(len(p.postponed)) {
-		p.write = 0
-	}
 }
 
-func (p *retryMsgWorker) isEmptyPostponed() bool {
-	return p.read == p.write
+func (p *retryMsgWorker) processPostponed(pmsg postponedMsg) {
+	switch {
+	case p.processMsg(pmsg.msg, pmsg.repeatFn):
+	case p.postponedLimit > 1:
+		p.postponedLimit--
+	default:
+		p.postponedLimit = 0
```

Comment:
Here we can be only if p.postponedLimit already 0, it should be 0, less impossible. Why we need set here?

Reply:
can get here when postponedLimit == 1

Reply:
but yes, it can be changed as
	case p.postponedLimit >= 1:
		p.postponedLimit--
    }

Reply:
My bad